File: CSharp/Specs/BuildersTest.cs
using System;
using NUnit.Framework;
using ThingModel.Builders;

namespace ThingModel.Specs
{
	[TestFixture]
	class BuildersTest
	{
		[Test]
		public void TestThingTypeBuilder()
		{
			ThingType type = BuildANewThingType.Named("rabbit")
				.WhichIs("Just a rabbit")
				.ContainingA.String("name")
				.AndA.LocationLatLng("localization")
				.AndA.NotRequired.Double("speed")
				.AndAn.Int("nbChildren", "Number of children")
					.WhichIs("The number of direct children for the rabbit");

			Assert.That(type, Is.Not.Null);
			Assert.That(type.Description, Is.EqualTo("Just a rabbit"));

			Assert.That(type.GetPropertyDefinition("name"), Is.Not.Null);
			Assert.That(type.GetPropertyDefinition("speed"), Is.Not.Null);
			Assert.That(type.GetPropertyDefinition("speed").Required, Is.False);
			Assert.That(type.GetPropertyDefinition("nbChildren").Required, Is.True);
			Assert.That(type.GetPropertyDefinition("nbChildren").Description, Contains.Substring("direct children"));

			ThingType superRabbit = BuildANewThingType.Named("super_rabbit")
				.WhichIs("A better rabbit")
				.ContainingAn.Int("power")
				.AndA.CopyOf(type);

			Assert.That(superRabbit.GetPropertyDefinition("name"), Is.Not.Null);
			Assert.That(superRabbit.GetPropertyDefinition("power"), Is.Not.Null);

			superRabbit.GetPropertyDefinition("nbChildren").Name = "Number of super children";
			
			Assert.That(type.GetPropertyDefinition("nbChildren").Name, Does.Not.Contain("super"));
		}

		[Test]
		public void TestThingBuilder()
		{
			ThingType duckType = BuildANewThingType.Named("duck");

			Thing duck = BuildANewThing.As(duckType)
				.IdentifiedBy("ab548")
				.ContainingA.String("name", "Roger")
                .AndA.Location(new Location.Point())
				.AndA.Location("destination", new Location.Point())
				.AndAn.Int("nbChildren", 12);

			Assert.That(duck.ID, Is.EqualTo("ab548"));
			Assert.That(duck.Type.Name, Is.EqualTo("duck"));
			Assert.That(duck.GetProperty<Property.Int>("nbChildren").Value,
				Is.EqualTo(12));

            Thing spaceShip = BuildANewThing.WithoutType()
                .IdentifiedBy("spaceShip")
                .ContainingAn.Int("speed", 12)
                .AndA.Location(new Location.LatLng(1, 2, 3))
                .AndA.Location(new Location.Equatorial(1, 2, 3)) // change type
                .AndA.Location("destination", new Location.Equatorial(4, 5, 6))
                .AndA.Location("from", new Location.LatLng(7,8))
                .AndA.DateTime("arrivalTime", new DateTime(2087, 8, 14))
                .AndA.Double("averageSpeed", Double.NaN);
			
            Assert.That(spaceShip.ID, Is.EqualTo("spaceShip"));
		    Assert.That(spaceShip.Type, Is.Null);
			Assert.That(spaceShip.LocationEquatorial().Declination, Is.EqualTo(2));
			Assert.That(spaceShip.LocationLatLng("from").Latitude, Is.EqualTo(7));
			Assert.That(spaceShip.DateTime("arrivalTime"), Is.Not.Null);
		}

	    [Test]
	    public void TestBuilderLikeThingSyntax()
	    {
			ThingType duckType = new ThingType("duck"); 

			Thing duck = new Thing("ab548", duckType);
            duck.String("name", "Roger");
            duck.LocationPoint("location", new Location.Point());
            duck.LocationPoint("destination", new Location.Point());
            duck.Int("nbChildren", 12);

			Assert.That(duck.ID, Is.EqualTo("ab548"));
			Assert.That(duck.Type.Name, Is.EqualTo("duck"));
			Assert.That(duck.Int("nbChildren"),	Is.EqualTo(12));
			Assert.That(duck.LocationPoint().X, Is.EqualTo(0));
			Assert.That(duck.LocationPoint("destination").X, Is.EqualTo(0));

            Thing spaceShip = new Thing("spaceShip");
            spaceShip.ContainingAn.Int("speed", 12);
            spaceShip.LocationLatLng("location", new Location.LatLng(1, 2, 3));
            spaceShip.LocationEquatorial("location", new Location.Equatorial(1, 2, 3)); // change type
            spaceShip.LocationEquatorial("destination", new Location.Equatorial(4, 5, 6));
            
            spaceShip.ContainingA.Location("from", new Location.LatLng(7,8))
                .AndA.DateTime("arrivalTime", new DateTime(2087, 8, 14))
                .AndA.Double("averageSpeed", Double.NaN);
            
            Assert.That(spaceShip.ID, Is.EqualTo("spaceShip"));
		    Assert.That(spaceShip.Type, Is.Null);
			Assert.That(spaceShip.LocationEquatorial().Declination, Is.EqualTo(2));
			Assert.That(spaceShip.LocationLatLng("from").Latitude, Is.EqualTo(7));
			Assert.That(spaceShip.DateTime("arrivalTime"), Is.Not.Null);
	    }
	}
}

File: TypeScript/Specs/BuildersTest.ts
/// <reference path="../bower_components/DefinitelyTyped/mocha/mocha.d.ts" />
/// <reference path="../bower_components/DefinitelyTyped/should/should.d.ts" />

import should = require('should');


describe("Thing test", () => {

	beforeEach(() => {

	});

	// ReSharper disable WrongExpressionStatement
	describe("testThingTypeBuilder", () => {
		var type = ThingModel.BuildANewThingType.Named("rabbit")
			.WhichIs("Just a rabbit")
			.ContainingA.String("name")
			.AndA.LocationLatLng()
			.AndA.NotRequired.Double("speed")
			.AndAn.Int("nbChildren", "Number of children")
			.WhichIs("The number of direct children")
			.Build();

		it('should contains the properties', () => {
			(type == null).should.be.false;

			type.GetPropertyDefinition("name").should.be.ok;
			type.GetPropertyDefinition("speed").should.be.ok;
			type.GetPropertyDefinition("speed").Required.should.be.false;
			type.GetPropertyDefinition("nbChildren").Required.should.be.true;
			type.GetPropertyDefinition("nbChildren").Description.should.match(/direct children/);
		});

		var superRappit = ThingModel.BuildANewThingType.Named("super_rabbit")
			.WhichIs("A better rabbit")
			.ContainingA.CopyOf(type)
			.AndAn.Int("power")
			.Build();

		it('should create a copy of the type', () => {
			superRappit.GetPropertyDefinition("name").should.be.ok;
			superRappit.GetPropertyDefinition("power").should.be.ok;

			superRappit.GetPropertyDefinition("nbChildren").Name = "Number of super children";
			type.GetPropertyDefinition("nbChildren").Name.should.not.match(/super/);
		});
	});


	describe("test thing builder", () => {
		var duckType = ThingModel.BuildANewThingType.Named("duck").Build();

		var duck = ThingModel.BuildANewThing.As(duckType)
			.IdentifiedBy("ab548")
			.ContainingA.String("name", "Roger")
			.AndA.Location(new ThingModel.Location.Point())
			.AndAn.Int("nbChildren", 12)
			.Build();

		it("should contains something", () => {
			duck.ID.should.be.equal("ab548");
			duck.Type.Name.should.be.equal("duck");
			duck.Int("nbChildren").should.be.equal(12);
		});
	});

	describe("BootStrap", () => {
		var thing = ThingModel.BuildANewThingType.Named("Thing")
			.ContainingA.String("id")
			.AndA.NotRequired.String("typename").Build();

		var thingPropierties = ThingModel.BuildANewThingType.Named("Property")
			.ContainingA.String("key")
			.AndA.NotRequired.String("value_string")
			.AndA.NotRequired.Int("value_int")
			.AndA.NotRequired.Double("value_double")
			.AndA.NotRequired.Boolean("value_boolean")
			.AndA.NotRequired.DateTime("value_datetime")
			.AndA.NotRequired.LocationLatLng("value_location").Build();

		var thingType = ThingModel.BuildANewThingType.Named("Thing type")
			.ContainingA.NotRequired.String("name")
			.AndA.NotRequired.String("description").Build();

		var propertyType = ThingModel.BuildANewThingType.Named("Property type")
			.ContainingA.String("key")
			.AndA.NotRequired.String("name")
			.AndA.NotRequired.String("description")
			.AndA.Boolean("required")
			.AndA.String("type").Build();

		// thingProperties mustt contain one of the notrequired properties
		// propertyType type should be a value in string/location/double/int/boolean/datetime
		// the properties should be connected to their things
		// the properties types should be connected to their things type
	});
});